README.md

```diff
@@ -1,2 +1,277 @@
+    *   Return the output stream of a job a specific worker (with worker ID and job ID)
+
+
+# Controller Overview
+
+* [Controller Details](Server/Controller/README) to be placed here. 
+* # Controller
+
+Controller exposes an HTTP-server that gives a gateway to the controller-worker-cluster. HTTP requests to the controller api are made to workers on behalf of client cli.
+
+- controller runs an HTTP server as well as a GRPC server.
+- controller keeps record of the available workers in the cluster.
+- Workers can register and deregister on the controller.
+
+
+## Data Structures
+---
+
+# Workers
+
+Workers exposes a GRPC-server to communicate with the controller. Main job of workers are to run specific jobs requested by the controller and report on those jobs. Jobs can be ruby/python/bash scripts or any executables available on the worker machine.
+
+## Data Structures
+---
+
+### Jobs
+
+```golang
+// jobsMutex is the lock to access jobs map.
+// jobs is the map that holds current/past jobs.
+//    - key: job id
+//    - value: pointer to the created job object.
+var (
+	jobsMutex = &sync.Mutex{}
+	jobs      = make(map[string]*job)
+)
+
+// job holds information about the ongoing or past jobs,
+// that were triggered by the controller.
+//    - id: UUID assigned by the worker and sent back to the controller.
+//    - command: command which the controller run the job with
+//    - path: path to the job file/executable sent by the controller.
+//    - outFilePath: file path to where the output of the job will be piped.
+//    - cmd: pointer to the cmd.Exec command to get job status etc.
+//    - done: whether if job is done (default false)
+//    - err: error while running the job (default nil)
+type job struct {
+	id          string
+	command     string
+	path        string
+	outFilePath string
+	cmd         *exec.Cmd
+	done        bool
+	err         error
+}
+```
+
+## Congroller - Worker(s) Configuration
+---
+
+[grpc_server]
+addr = "${{grpc_server_addr}}"
+use_tls =  ${{use_tls}}
+crt_file = ${{crt_file}}
+key_file = ${{key_file}}
+
+[controller]
+addr = ${{controller_addr}}
+```
+
+`grpc_server`:
+  - `addr`: Address on which the GRPC server will be run.
+  - `use_tls`: Whether the GRPC server should use TLS. If `true`, `crt_file` and `key_file` should be provided.
+  - `crt_file`: Path to the certificate file for TLS.
+  - `key_file`: Path to the key file for TLS.
+
+`controller`:
+  - `address`: Address on which the GRPC server of the controller is run.
+
+## GRPC Server
+---
+
+#### Methods:
+```
+service Worker {
+  rpc StartJob(StartJobRequest) returns (StartJobResponse) {}
+  rpc StopJob(StopJobRequest) returns (StopJobResponse) {}
+  rpc QueryJob(QueryJobRequest) returns (QueryJobResponse) {}
+  rpc StreamJob(StreamJobRequest) returns (stream StreamJobResponse) {}
+}
+```
+
+### Variables:
+```
+message StartJobRequest {
+  string command = 1;
+  string path = 2;
+}
```

Comment:
Do we need information here regarding the amount of resources required for the job?

Reply:
I was planning on having that be part of the command string. But it could be added as request parameters